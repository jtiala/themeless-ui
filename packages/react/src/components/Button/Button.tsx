import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { ButtonHTMLAttributes, ReactNode } from "react";

type HTMLButtonProps = ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  /**
   * Type of the button.
   */
  type?: HTMLButtonProps["type"];

  /**
   * Name of the button.
   */
  name?: HTMLButtonProps["name"];

  /**
   * Prevents the user from interacting with the button: it cannot be pressed or focused.
   */
  disabled?: HTMLButtonProps["disabled"];

  /**
   * Callback to call when the button is clicked.
   */
  onClick?: HTMLButtonProps["onClick"];

  /**
   * Callback to call when the button receives focus.
   */
  onFocus?: HTMLButtonProps["onFocus"];

  /**
   * Callback to call when the button loses focus.
   */
  onBlur?: HTMLButtonProps["onBlur"];

  /**
   * Button content.
   */
  children: ReactNode;
} & CommonComponentProps;

const className = cn("button");

/**
 * A clickable UI element commonly used to trigger actions, such as submitting a form.
 */
export function Button({
  type = "button",
  name,
  disabled,
  onClick,
  onFocus,
  onBlur,
  children,
  id,
  testId,
}: ButtonProps) {
  return (
    <button
      type={type}
      name={name}
      disabled={disabled}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      className={className}
      id={id}
      data-testid={testId}
    >
      {children}
    </button>
  );
}
