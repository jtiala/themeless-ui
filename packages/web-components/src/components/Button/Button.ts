import style from "@themeless-ui/style/dist/button.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import {
  TUIBlurEvent,
  TUIClickEvent,
  TUIFocusEvent,
  TUIMouseEnterEvent,
  TUIMouseLeaveEvent,
  TUITouchEndEvent,
  TUITouchStartEvent,
} from "../../events";
import { TUIComponent } from "../TUIComponent";

const className = cn("button");

@customElement("tui-button")
export class Button extends TUIComponent {
  /**
   * Type of the button.
   */
  @property()
  type?: HTMLButtonElement["type"];

  /**
   * Name of the button.
   */
  @property()
  name?: HTMLButtonElement["name"];

  /**
   * Prevents the user from interacting with the button: it cannot be pressed or focused.
   */
  @property({ type: Boolean })
  disabled?: HTMLButtonElement["disabled"];

  static styles = unsafeCSS(style);

  constructor() {
    super();

    this.type = "button";
  }

  private _handleClick() {
    this.dispatchEvent(new TUIClickEvent());
  }

  private _handleFocus() {
    this.dispatchEvent(new TUIFocusEvent());
  }
  private _handleBlur() {
    this.dispatchEvent(new TUIBlurEvent());
  }

  private _handleMouseEnter() {
    this.dispatchEvent(new TUIMouseEnterEvent());
  }

  private _handleMouseLeave() {
    this.dispatchEvent(new TUIMouseLeaveEvent());
  }

  private _handleTouchStart() {
    this.dispatchEvent(new TUITouchStartEvent());
  }

  private _handleTouchEnd() {
    this.dispatchEvent(new TUITouchEndEvent());
  }

  override render() {
    return html`
      <button
        type="${this.type}"
        name="${this.name || nothing}"
        ?disabled="${this.disabled}"
        @click="${this._handleClick}"
        @focus="${this._handleFocus}"
        @blur="${this._handleBlur}"
        @mouseEnter="${this._handleMouseEnter}"
        @mouseLeave="${this._handleMouseLeave}"
        @touchStart="${this._handleTouchStart}"
        @touchEnd="${this._handleTouchEnd}"
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      >
        <slot></slot>
      </button>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-button": Button;
  }
}
