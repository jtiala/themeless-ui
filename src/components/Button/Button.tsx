import { ButtonHTMLAttributes, ReactNode } from "react";

type HTMLButtonProps = ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  /**
   * Button content.
   */
  children: ReactNode;

  /**
   * Type of the button.
   */
  type?: HTMLButtonProps["type"];

  /**
   * Callback to call when the button is clicked.
   */
  onClick?: HTMLButtonProps["onClick"];

  /**
   * Callback to call when the button receives focus.
   */
  onFocus?: HTMLButtonProps["onFocus"];

  /**
   * Callback to call when the button loses focus.
   */
  onBlur?: HTMLButtonProps["onBlur"];
};

/**
 * A clickable UI element commonly used to trigger actions, such as submitting a form.
 */
export function Button({ type = "button", ...rest }: ButtonProps) {
  return <button type={type} {...rest} />;
}
