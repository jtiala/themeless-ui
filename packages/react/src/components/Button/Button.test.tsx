import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
  userEvent,
  vi,
} from "../../../test";
import { Button } from "./Button";

const onClick = vi.fn();
const onFocus = vi.fn();
const onBlur = vi.fn();

const defaultButton = (
  <Button
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="button"
  >
    Button
  </Button>
);

const disabledButton = (
  <Button
    disabled
    onClick={() => onClick()}
    onFocus={() => onFocus()}
    onBlur={() => onBlur()}
    testId="button"
  >
    Button
  </Button>
);

describe("Button", async () => {
  it("should render the button", () => {
    expect(componentToJson(renderer.create(defaultButton))).toMatchSnapshot();

    render(defaultButton);

    expect(screen.getByTestId("button")).toBeInTheDocument();
  });

  it("should call the callbacks", async () => {
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(defaultButton);
    const button = screen.getByTestId("button");

    expect(button).toBeInTheDocument();
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(button);
    expect(onClick.mock.calls.length).toBe(1);
    expect(onFocus.mock.calls.length).toBe(1);

    button.blur();
    expect(onBlur.mock.calls.length).toBe(1);

    button.focus();
    expect(onFocus.mock.calls.length).toBe(2);

    await userEvent.type(button, "{space}");
    expect(onClick.mock.calls.length).toBe(2);
  });

  it("shouldn't be possible to access a disabled button", async () => {
    onClick.mockReset();
    onFocus.mockReset();
    onBlur.mockReset();

    render(disabledButton);
    const button = screen.getByTestId("button");

    expect(button).toBeInTheDocument();
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);
    expect(onBlur.mock.calls.length).toBe(0);

    await userEvent.click(button);
    expect(onClick.mock.calls.length).toBe(0);
    expect(onFocus.mock.calls.length).toBe(0);

    button.blur();
    expect(onBlur.mock.calls.length).toBe(0);

    button.focus();
    expect(onFocus.mock.calls.length).toBe(0);

    await userEvent.type(button, "{space}");
    expect(onClick.mock.calls.length).toBe(0);
  });
});
