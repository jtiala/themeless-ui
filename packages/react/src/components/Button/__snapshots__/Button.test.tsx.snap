// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match the snapshot 1`] = `
<button
  className="tui__button"
  data-testid="button"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  type="button"
>
  Button
</button>
`;
